tower-posix/src/Ivory/OS/Posix/Tower/Pthread.hs

```diff
@@ -0,0 +1,574 @@
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE QuasiQuotes #-}
+{-# LANGUAGE TypeOperators #-}
+{-# OPTIONS_GHC -fno-warn-orphans #-}
+
+module Ivory.OS.Posix.Tower.Pthread where
+
+import Ivory.Language
+import Ivory.Language.Type
+import Ivory.Artifact
+import qualified Paths_tower_posix
+
+pthread_header :: String
+pthread_header = "pthread.h"
+
+sched_header :: String
+sched_header = "sched.h"
+
+ivory_sched_header :: String
+ivory_sched_header = "ivory_sched.h"
+
+-------------
+-- TYPES FOR 32 BITS
+-------------
+
+type CSize_t = Uint32
+type CInt    = Sint32
+
+-- Void type
+type VoidType =  NewType "void"
+
+instance IvoryType VoidType where
+  ivoryType _ = mkNewType (Proxy :: Proxy "void")
+
+-- other structs impoted in pthread.h
+
+[ivory|
+abstract struct sched_param "sched.h"
+abstract struct sched_attr "sched.h"
+abstract struct timespec "time.h"
+|]
+
+-- Pthread types, defined in sys/types.h
```

Comment:
do you think it'd be worth adding new concrete syntax for this? if you haven't played with Template Haskell before this might be a good opportunity to try it out since the pattern here is so regular